proposed/0000-UDP-hole-punching.md

```diff
@@ -0,0 +1,191 @@
+to allow for P2P connections of nodes behind NATs in a fully decentralised
+network.
+
+# Motivation
+
+Most users of the Crust library are expected to run the software on home PCs connected
+to the Internet through a router. Routers usually implement some kind of a firewall to
+prevent anyone from the outside Internet being able to connect to the user's PC directly.
+As Crust is a library that creates P2P connections, it should try available techniques
+to circumvent these firewall restrictions and UDP hole punching is one such
+technique.
+
+Additionally, the traditional UDP hole punching process involves a well known
+server to tell each peer their external UDP endpoints. Since our intent
+is a fully decentralised system, this text also describes how nodes to which
+we're already connected can be used as a replacement for such a server.
+
+In the following text the reader is expected to know basic NAT types and how they work.
+The section 'Methods of translation' on [wikipedia](https://en.wikipedia.org/wiki/Network_address_translation)
+gives a nice overview.
+
+# Detailed design
+
+The UDP hole punching process described here consists of two steps:
+
+1. Finding an external mapping of a UDP socket
+2. Do the actual hole punching.
+
+## Finding an external mapping of a UDP socket
+
+Suppose, we have two nodes `A` and `B`, where each one is possibly behind a NAT and
+they want to connect to each other.  Routing needs to send `A` an endpoint of `B` and
+vice versa. But `A` and `B` don’t know their public
+endpoints a priori (they may know their IP addresses from previous connection establishments,
+but not their ports). Each node therefore needs to create a UDP socket and use it to ask some
+other node it is connected to what the public endpoint of the UDP socket is. We’ll refer
+to such helping nodes `C(A)` and `C(B)` (it need not be the same node for `A` and `B`).
+We'll call `U(C(X))` the UDP port on which X can contact `C(X)` in order to find its public
+endpoint. `X` finds out about `U(C(X))` through the initial handhsake, that is, when
+`X` connects to `C(X)`.
+
+In the following text we’ll describe steps taken at node `A`, but the same steps need to
+be taken at node `B`.
+
+For `A` to find out its external address `A` first needs to create a UDP socket `S(A)`
+with an arbitrary local port.  `A` will then use `S(A)` to periodically send
+datagrams to `U(C(A))` containing a MAGIC number and the ID of the request. Each time
+`C(A)` receives such datagram, it sends back a datagram containing public endpoint
+of `S(A)` together with the request ID.
+
+The above protocol shall be initiated from upper layers by calling
+
+    Service::get_mapped_udp_socket(callback: FnBox<(io::Result<MappedUdpSocket>)>)
+
+And the result of such a call shall be an invocation of the callback.
+
+where `MappedUdpSocket` is defined as
+
+    struct MappedUdpSocket {
+        udp_socket: UdpSocket,
+        public_address: SocketAddr, // of node A
+    }
+
```

Comment:
The client (Routing in our own case) needs to pass a token down to Crust when making this call so that it can be passed back in the `MappedUdpSocket` I believe.

If the client were to pass down the public encryption key (`::sodiumoxide::crypto::box_::PublicKey`) of the peer it's trying to connect to via this new socket, then not only would it serve the purpose of being a token for the client, but it could also be used to encrypt *all* communications with the peer - from handshaking onwards.

If this were the case, we could also do away with the `secret` field in the hole-punching functions/structs if we wanted.

I'm not suggesting that we actually implement the encryption as part of this RFC; just that if we need to receive a token from Routing anyway, we might as well make it a public key which is unused by Crust for now other than to pass back to Routing once the connection is made.